src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_cython.pyx

```diff
@@ -526,62 +510,121 @@ cdef class PyDBFrame:
+        # if DEBUG: print('frame trace_dispatch %s %s %s %s %s %s, stop: %s' % (frame.f_lineno, frame.f_code.co_name, frame.f_code.co_filename, event, constant_to_str(info.pydev_step_cmd), arg, info.pydev_step_stop))
         try:
-            info.is_tracing = True
+            info.is_tracing += 1
             line = frame.f_lineno
             line_cache_key = (frame_cache_key, line)
 
             if main_debugger.pydb_disposed:
                 return None if event == 'call' else NO_FTRACE
 
             plugin_manager = main_debugger.plugin
-
+            is_coroutine_or_generator = frame.f_code.co_flags & 0xa0  # 0xa0 ==  CO_GENERATOR = 0x20 | CO_COROUTINE = 0x80
             is_exception_event = event == 'exception'
             has_exception_breakpoints = main_debugger.break_on_caught_exceptions or main_debugger.has_plugin_exception_breaks
 
-            if is_exception_event:
-                if has_exception_breakpoints:
-                    should_stop, frame = self.should_stop_on_exception(frame, event, arg)
-                    if should_stop:
-                        self.handle_exception(frame, event, arg)
+            stop_frame = info.pydev_step_stop
+            step_cmd = info.pydev_step_cmd
+            if is_coroutine_or_generator:
+                # Dealing with coroutines and generators:
+                # When in a coroutine we change the perceived event to the debugger because
+                # a call, StopIteration exception and return are usually just pausing/unpausing it.
+                if event == 'line':
+                    is_line = True
+                    is_call = False
+                    is_return = False
+
+                elif event == 'return':
+                    is_line = False
+                    is_call = False
+                    is_return = True
+
+                    returns_cache_key = (frame_cache_key, 'returns')
+                    return_lines = frame_skips_cache.get(returns_cache_key)
+                    if return_lines is None:
+                        # Note: we're collecting the return lines by inspecting the bytecode as
+                        # there are multiple returns and multiple stop iterations when awaiting and
+                        # it doesn't give any clear indication when a coroutine or generator is
+                        # finishing or just pausing.
+                        return_lines = set()
+                        for x in main_debugger.collect_return_info(frame.f_code):
+                            # Note: cython does not support closures in cpdefs (so we can't use
+                            # a list comprehension).
+                            return_lines.add(x.return_line)
+
+                        frame_skips_cache[returns_cache_key] = return_lines
+
+                    if line not in return_lines:
+                        # Not really a return (coroutine/generator paused).
```

Comment:
I wonder if this should perhaps be special-cased for "step in", so that it's possible to step into a yield, and land at the point where `next()` is being called? It can be handy sometimes. 

Not sure whether the equivalent for `await` would make much sense, though...

Reply:
Right now if the user does a step in (instead of a step over), he'll get to the point where the `next()` is being called (which is what I want to fix with asyncio as with generators in general that's where the user wants to stop, but with coroutines I don't think he ever wants to get into the asyncio code).

I agree that for generators in general we should revert to the usual step in (which would get to the `next()` being called). My plan is fixing that soon (when I also fix the step in for coroutines).